File: providers/replicate.py
"""Replicate API provider — Hailuo 2.3, Wan 2.2 T2V, Wan 2.2 I2V."""

import asyncio
import time

import httpx

from .base import API_KEYS

REPLICATE_API = "https://api.replicate.com/v1"


def _build_hailuo_input(prompt: str, params: dict) -> dict:
    """Build Replicate input payload for MiniMax Hailuo 2.3."""
    duration = params.get("duration", 6)
    resolution = params.get("resolution", "768p")
    image = params.get("image_data_uri")

    # Snap duration to valid values
    duration = 10 if duration >= 8 else 6
    # Validate resolution
    if resolution not in ("768p", "1080p"):
        resolution = "768p"
    # 1080p only supports 6s
    if resolution == "1080p":
        duration = 6

    inp: dict = {
        "prompt": prompt,
        "duration": duration,
        "resolution": resolution,
        "prompt_optimizer": params.get("optimize_prompt", True),
    }
    if image:
        inp["first_frame_image"] = image
    return inp


def _build_wan_t2v_input(prompt: str, params: dict) -> dict:
    """Build Replicate input payload for Wan 2.2 14B T2V Fast."""
    inp: dict = {
        "prompt": prompt,
        "aspect_ratio": params.get("aspect_ratio", "16:9"),
        "resolution": params.get("resolution", "480p"),
        "num_frames": params.get("num_frames", 81),
        "frames_per_second": params.get("frames_per_second", 16),
        "sample_shift": params.get("sample_shift", 12),
        "go_fast": params.get("go_fast", True),
        "interpolate_output": params.get("interpolate_output", True),
    }
    negative_prompt = params.get("negative_prompt")
    if negative_prompt:
        inp["negative_prompt"] = negative_prompt
    # Optional LoRA
    lora = params.get("lora_weights_transformer")
    if lora:
        inp["lora_weights_transformer"] = lora
        inp["lora_scale_transformer"] = params.get("lora_scale_transformer", 1)
    lora2 = params.get("lora_weights_transformer_2")
    if lora2:
        inp["lora_weights_transformer_2"] = lora2
        inp["lora_scale_transformer_2"] = params.get("lora_scale_transformer_2", 1)
    return inp


def _build_wan_i2v_input(prompt: str, params: dict) -> dict:
    """Build Replicate input payload for Wan 2.2 14B I2V."""
    image = params.get("image_data_uri")
    if not image:
        raise ValueError("Wan I2V requires an image — upload one before generating.")

    inp: dict = {
        "prompt": prompt,
        "image": image,
        "resolution": params.get("resolution", "480p"),
        "num_frames": min(params.get("num_frames", 81), 100),
        "frames_per_second": min(params.get("frames_per_second", 16), 24),
        "sample_steps": params.get("sample_steps", 40),
        "sample_shift": params.get("sample_shift", 5),
        "go_fast": params.get("go_fast", False),
    }
    negative_prompt = params.get("negative_prompt")
    if negative_prompt:
        inp["negative_prompt"] = negative_prompt
    return inp


def _build_pvideo_input(prompt: str, params: dict) -> dict:
    """Build Replicate input payload for PrunaAI P-Video.

    Two variants share this builder, dispatched via params["variant"]:
      - "landscape": locks aspect_ratio="16:9" so the multi-crop pipeline
        (FORCE_LANDSCAPE) can slice into 9:16 clips, mirroring Hailuo's UX.
      - "vertical": locks aspect_ratio="9:16" for native vertical output,
        single clip per generation.
    """
    variant = params.get("variant", "landscape")
    aspect_ratio = "9:16" if variant == "vertical" else "16:9"

    duration = max(1, min(int(params.get("duration", 6)), 10))
    resolution = params.get("resolution", "720p")
    if resolution not in ("720p", "1080p"):
        resolution = "720p"

    return {
        "prompt": prompt,
        "duration": duration,
        "aspect_ratio": aspect_ratio,
        "resolution": resolution,
        "fps": 24,
        "draft": False,
        "prompt_upsampling": params.get("optimize_prompt", True),
        "disable_safety_filter": True,
        "save_audio": False,
        "no_op": False,
    }


def _build_wan_i2v_fast_input(prompt: str, params: dict) -> dict:
    """Build Replicate input payload for Wan 2.2 14B I2V Fast (PrunaAI)."""
    image = params.get("image_data_uri")
    if not image:
        raise ValueError("Wan I2V Fast requires an image — upload one before generating.")

    inp: dict = {
        "prompt": prompt,
        "image": image,
        "resolution": params.get("resolution", "480p"),
        "num_frames": params.get("num_frames", 81),
        "frames_per_second": params.get("frames_per_second", 16),
        "sample_shift": params.get("sample_shift", 12),
        "go_fast": params.get("go_fast", True),
        "interpolate_output": params.get("interpolate_output", False),
    }
    negative_prompt = params.get("negative_prompt")
    if negative_prompt:
        inp["negative_prompt"] = negative_prompt
    # Optional last frame
    last_image = params.get("last_image_data_uri")
    if last_image:
        inp["last_image"] = last_image
    # Optional LoRA
    lora = params.get("lora_weights_transformer")
    if lora:
        inp["lora_weights_transformer"] = lora
        inp["lora_scale_transformer"] = params.get("lora_scale_transformer", 1)
    lora2 = params.get("lora_weights_transformer_2")
    if lora2:
        inp["lora_weights_transformer_2"] = lora2
        inp["lora_scale_transformer_2"] = params.get("lora_scale_transformer_2", 1)
    return inp


_INPUT_BUILDERS = {
    "minimax/hailuo-2.3": _build_hailuo_input,
    "wan-video/wan-2.2-t2v-fast": _build_wan_t2v_input,
    "wan-video/wan-2.2-i2v-a14b": _build_wan_i2v_input,
    "wan-video/wan-2.2-i2v-fast": _build_wan_i2v_fast_input,
    "prunaai/p-video": _build_pvideo_input,
}


# ---------------------------------------------------------------------------
# Transient-failure handling for video generation.
#
# Production evidence (2026-09-23/24): paid-for Replicate predictions were
# abandoned because one poll GET hit a ReadError/ConnectTimeout, and whole
# replenishment jobs failed on a 429 throttle or a 5xx at submission.  Each
# such job blocked its page's refill for hours downstream.  Only failures that
# cannot create a second paid prediction are retried here:
#
# * submission: HTTP 429 (honouring ``retry_after``), 500 and 503 -- Replicate
#   returned no prediction id in every observed case (2026-09-24) -- or a
#   connection that was never established.  502/504 are gateway results whose
#   upstream may already have created the prediction, so they are NOT retried.  A read/write failure after the
#   request body may have reached Replicate is ambiguous and is NOT retried,
#   so a lost response can never become a duplicate paid prediction.
# * polling: the prediction already exists, so a transport error, 429, 5xx or
#   unparseable body is retried on the SAME prediction until the deadline.
# * "Prediction interrupted; please retry (code: PA)" is resubmitted once.
#
# Insufficient credit (402), validation errors, provider-side failures and
# timeouts are never retried.  A prediction that exceeds the deadline is
# cancelled so output that will be discarded stops accruing cost.  The model,
# input and prompt are identical on every attempt; nothing here changes the
# provider a recipe pinned.
# ---------------------------------------------------------------------------
START_ATTEMPTS = 4
START_RETRY_STATUSES = frozenset({429, 500, 503})
START_BACKOFF_SECONDS = (2.0, 5.0, 10.0)
RETRY_AFTER_CAP_SECONDS = 30.0
POLL_INTERVAL_SECONDS = 5.0
POLL_TRANSIENT_LIMIT = 12
PREDICTION_DEADLINE_SECONDS = 600.0
INTERRUPTED_RESUBMITS = 1
_INTERRUPTED_MARKER = "(code: PA)"


async def _sleep(seconds: float) -> None:
    await asyncio.sleep(seconds)


def _now() -> float:
    return time.monotonic()


def _retry_after_seconds(resp: httpx.Response, attempt: int) -> float:
    fallback = START_BACKOFF_SECONDS[min(attempt, len(START_BACKOFF_SECONDS) - 1)]
    if resp.status_code != 429:
        return fallback
    try:
        hinted = float(resp.json().get("retry_after"))
    except Exception:
        try:
            hinted = float(resp.headers.get("retry-after", ""))
        except ValueError:
            return fallback
    if hinted != hinted or hinted <= 0:  # NaN or non-positive
        return fallback
    return min(hinted, RETRY_AFTER_CAP_SECONDS)


async def _start_prediction(
    client: httpx.AsyncClient, headers: dict, model_id: str, input_params: dict,
) -> str:
    for attempt in range(START_ATTEMPTS):
        final = attempt == START_ATTEMPTS - 1
        try:
            resp = await client.post(
                f"{REPLICATE_API}/models/{model_id}/predictions",
                headers=headers,
                json={"input": input_params},
                timeout=30,
            )
        except (httpx.ConnectError, httpx.ConnectTimeout):
            # No connection means no request reached Replicate: safe to retry.
            if final:
                raise
            await _sleep(START_BACKOFF_SECONDS[min(attempt, len(START_BACKOFF_SECONDS) - 1)])
            continue
        if resp.status_code in (200, 201):
            return resp.json()["id"]
        if resp.status_code in START_RETRY_STATUSES and not final:
            await _sleep(_retry_after_seconds(resp, attempt))
            continue
        raise RuntimeError(f"Replicate start failed: {resp.text}")
    raise RuntimeError("Replicate start failed: retries exhausted")  # pragma: no cover


async def _cancel_prediction(client: httpx.AsyncClient, headers: dict, pred_id: str) -> None:
    """Best effort: stop an abandoned prediction from producing billable output."""
    try:
        await client.post(
            f"{REPLICATE_API}/predictions/{pred_id}/cancel", headers=headers, timeout=15,
        )
    except Exception:
        pass


async def _await_prediction(
    client: httpx.AsyncClient, headers: dict, pred_id: str,
) -> tuple[str, object]:
    """Poll one existing prediction. Returns ("succeeded", output) or (status, error)."""
    poll_url = f"{REPLICATE_API}/predictions/{pred_id}"
    deadline = _now() + PREDICTION_DEADLINE_SECONDS
    transient = 0
    while _now() < deadline:
        data = None
        try:
            r = await client.get(poll_url, headers=headers, timeout=30)
            if r.status_code == 429 or r.status_code >= 500:
                raise httpx.HTTPStatusError("transient poll status", request=r.request, response=r)
            if r.status_code != 200:
                raise RuntimeError(f"Replicate poll failed: HTTP {r.status_code}: {r.text[:200]}")
            data = r.json()
            if not isinstance(data, dict):
                raise ValueError("poll body is not an object")
        except (httpx.TransportError, httpx.HTTPStatusError, ValueError) as error:
            transient += 1
            if transient > POLL_TRANSIENT_LIMIT:
                await _cancel_prediction(client, headers, pred_id)
                raise RuntimeError(
                    f"Replicate poll unavailable after {transient} attempts "
                    f"(prediction {pred_id}): {error!r}"
                ) from error
            await _sleep(POLL_INTERVAL_SECONDS)
            continue
        transient = 0
        status = data.get("status", "")
        if status == "succeeded":
            return status, data.get("output")
        if status in ("failed", "canceled"):
            return status, data.get("error") or data.get("logs") or "unknown error (no details from API)"
        await _sleep(POLL_INTERVAL_SECONDS)
    await _cancel_prediction(client, headers, pred_id)
    raise RuntimeError(
        f"Replicate generation timed out after {int(PREDICTION_DEADLINE_SECONDS)}s "
        f"(prediction {pred_id})"
    )


async def generate(prompt: str, params: dict, client: httpx.AsyncClient) -> str:
    """Submit a prediction to Replicate and poll until complete."""
    key = API_KEYS["replicate"]
    model_id = params["model_id"]
    entry = params["entry"]

    headers = {"Authorization": f"Token {key}", "Content-Type": "application/json"}

    builder = _INPUT_BUILDERS.get(model_id)
    if not builder:
        raise RuntimeError(f"No input builder for model: {model_id}")
    input_params = builder(prompt, params)

    for submission in range(INTERRUPTED_RESUBMITS + 1):
        pred_id = await _start_prediction(client, headers, model_id, input_params)
        entry["provider_request_id"] = pred_id
        entry["status"] = "polling"
        status, result = await _await_prediction(client, headers, pred_id)
        if status == "succeeded":
            if isinstance(result, str):
                return result
            if isinstance(result, list) and result:
                return result[0]
            raise RuntimeError(f"Replicate unexpected output: {result}")
        if (
            status == "failed"
            and _INTERRUPTED_MARKER in str(result)
            and submission < INTERRUPTED_RESUBMITS
        ):
            continue
        raise RuntimeError(f"Replicate {status}: {result}")
    raise RuntimeError("Replicate generation failed: resubmissions exhausted")  # pragma: no cover


async def _poll_prediction(
    client: httpx.AsyncClient,
    headers: dict,
    pred_id: str,
    label: str,
    timeout_secs: int = 180,
) -> str | list:
    """Poll a Replicate prediction until it completes. Returns the output."""
    poll_url = f"{REPLICATE_API}/predictions/{pred_id}"
    deadline = time.time() + timeout_secs

    while time.time() < deadline:
        r = await client.get(poll_url, headers=headers, timeout=30)
        data = r.json()
        status = data.get("status", "")
        if status == "succeeded":
            return data.get("output")
        if status in ("failed", "canceled"):
            raise RuntimeError(
                f"Replicate {label} {status}: {data.get('error', 'unknown')}"
            )
        await asyncio.sleep(3)

    raise RuntimeError(f"Replicate {label} timed out after {timeout_secs}s")


def _generate_text_mask(image_data_uri: str) -> str:
    """Generate a binary mask highlighting burned-in text regions.

    TikTok captions are typically white/bright text with dark stroke, positioned
    in the center-lower area. We detect these by:
    1. Convert to grayscale
    2. Threshold for bright pixels (text is white/near-white)
    3. Dilate to connect letter strokes into solid regions
    4. Return as a base64 PNG data URI (white = remove, black = keep)
    """
    import base64
    import io

    from PIL import Image, ImageFilter

    # Decode the data URI
    _, b64data = image_data_uri.split(",", 1)
    img = Image.open(io.BytesIO(base64.b64decode(b64data)))
    w, h = img.size

    # Convert to grayscale
    gray = img.convert("L")

    # Threshold: pixels brighter than 200 are likely text
    threshold = 200
    mask = gray.point(lambda p: 255 if p > threshold else 0, mode="1")

    # Convert back to L mode for filtering
    mask = mask.convert("L")

    # Dilate to connect strokes and cover text edges/stroke outlines
    # MaxFilter expands white regions — run multiple passes for solid coverage
    for _ in range(5):
        mask = mask.filter(ImageFilter.MaxFilter(5))

    # Only keep the center 80% horizontally
    # (TikTok captions are typically centered, not in corners)
    final_mask = Image.new("L", (w, h), 0)
    # Crop region: center 80% width, full height (captions can be anywhere vertically)
    margin_x = int(w * 0.10)
    final_mask.paste(mask.crop((margin_x, 0, w - margin_x, h)), (margin_x, 0))

    # Encode as PNG data URI
    buf = io.BytesIO()
    final_mask.save(buf, format="PNG")
    b64 = base64.b64encode(buf.getvalue()).decode()
    return f"data:image/png;base64,{b64}"


_LAMA_VERSION = "40e67426e1bf78199d78b36580389fbbdcb4c9cdc2bc2b489e99d713f167b3c5"


async def remove_text(
    image_data_uri: str, client: httpx.AsyncClient | None = None
) -> str:
    """Remove burned-in text from an image using local mask + LaMa inpainting.

    1. Locally generates a binary mask detecting bright text pixels (Pillow).
    2. Sends image + mask to LaMa (dpakkk/image-object-removal) for inpainting.

    Args:
        image_data_uri: Base64 data URI of the source image.
        client: Optional shared httpx client; one is created if not provided.

    Returns:
        URL of the cleaned image on success.

    Raises:
        ValueError: If image_data_uri is empty/falsy.
        RuntimeError: If API key is missing, the prediction fails, or it times out.
    """
    if not image_data_uri:
        raise ValueError("image_data_uri is required — provide a base64 data URI")

    key = API_KEYS.get("replicate")
    if not key:
        raise RuntimeError("REPLICATE_API_TOKEN not set")

    headers = {"Authorization": f"Token {key}", "Content-Type": "application/json"}
    owns_client = client is None
    if owns_client:
        client = httpx.AsyncClient()

    try:
        # Step 1: Generate text mask locally (fast, no API call)
        mask_data_uri = _generate_text_mask(image_data_uri)

        # Step 2: LaMa inpainting — community model, version-based endpoint
        resp = await client.post(
            f"{REPLICATE_API}/predictions",
            headers=headers,
            json={
                "version": _LAMA_VERSION,
                "input": {
                    "image": image_data_uri,
                    "mask": mask_data_uri,
                },
            },
            timeout=30,
        )
        if resp.status_code not in (200, 201):
            raise RuntimeError(f"Replicate LaMa start failed: {resp.text}")

        pred_id = resp.json()["id"]
        output = await _poll_prediction(client, headers, pred_id, "LaMa")

        if isinstance(output, str):
            return output
        if isinstance(output, list) and output:
            return output[0]
        raise RuntimeError(f"LaMa unexpected output: {output}")

    finally:
        if owns_client:
            await client.aclose()

File: rust/crates/api/src/providers/replicate.rs
//! Replicate adapter. Port of providers/replicate.py's submit→poll loop plus
//! the per-model input builders for the curated v1 models.

use std::time::Duration;

use anyhow::{anyhow, Context};
use serde_json::{json, Value};

use super::{truncate, GenParams, Provider};

const API: &str = "https://api.replicate.com/v1";

pub struct ReplicateProvider {
    key: String,
}

impl ReplicateProvider {
    pub fn new(key: String) -> Self {
        Self { key }
    }
}

type Extra = serde_json::Map<String, Value>;

// Typed readers over the model-specific `extra` map (frontend-supplied knobs),
// each falling back to the Python builder's default.
fn ex_i(ex: &Extra, k: &str, d: i64) -> i64 {
    ex.get(k).and_then(Value::as_i64).unwrap_or(d)
}
fn ex_f(ex: &Extra, k: &str, d: f64) -> f64 {
    ex.get(k).and_then(Value::as_f64).unwrap_or(d)
}
fn ex_b(ex: &Extra, k: &str, d: bool) -> bool {
    ex.get(k).and_then(Value::as_bool).unwrap_or(d)
}
fn ex_s<'a>(ex: &'a Extra, k: &str) -> Option<&'a str> {
    ex.get(k).and_then(Value::as_str).filter(|s| !s.is_empty())
}

/// Build the model-specific `input` object. Faithful port of the Python
/// `_build_*_input` fns (providers/replicate.py) — reads the frontend's
/// model-specific knobs from `p.extra`, falling back to the same defaults;
/// rejects models we didn't wire up.
fn build_input(model_id: &str, p: &GenParams) -> anyhow::Result<Value> {
    let ex = &p.extra;
    match model_id {
        "minimax/hailuo-2.3" => {
            // Snap duration to {6,10}; 1080p forces 6s.
            let mut duration = if p.duration >= 8 { 10 } else { 6 };
            let mut resolution = p.resolution.clone();
            if resolution != "768p" && resolution != "1080p" {
                resolution = "768p".into();
            }
            if resolution == "1080p" {
                duration = 6;
            }
            let mut inp = json!({
                "prompt": p.prompt,
                "duration": duration,
                "resolution": resolution,
                "prompt_optimizer": ex_b(ex, "optimize_prompt", true),
            });
            if let Some(img) = &p.image_data_uri {
                inp["first_frame_image"] = json!(img);
            }
            Ok(inp)
        }
        "wan-video/wan-2.2-t2v-fast" => {
            let mut inp = json!({
                "prompt": p.prompt,
                "aspect_ratio": if p.aspect_ratio.is_empty() { "16:9".into() } else { p.aspect_ratio.clone() },
                "resolution": if p.resolution.is_empty() { "480p".into() } else { p.resolution.clone() },
                "num_frames": ex_i(ex, "num_frames", 81),
                "frames_per_second": ex_i(ex, "frames_per_second", 16),
                "sample_shift": ex_i(ex, "sample_shift", 12),
                "go_fast": ex_b(ex, "go_fast", true),
                "interpolate_output": ex_b(ex, "interpolate_output", true),
            });
            if let Some(np) = ex_s(ex, "negative_prompt") {
                inp["negative_prompt"] = json!(np);
            }
            if let Some(lora) = ex_s(ex, "lora_weights_transformer") {
                inp["lora_weights_transformer"] = json!(lora);
                inp["lora_scale_transformer"] = json!(ex_f(ex, "lora_scale_transformer", 1.0));
            }
            if let Some(lora2) = ex_s(ex, "lora_weights_transformer_2") {
                inp["lora_weights_transformer_2"] = json!(lora2);
                inp["lora_scale_transformer_2"] = json!(ex_f(ex, "lora_scale_transformer_2", 1.0));
            }
            Ok(inp)
        }
        "wan-video/wan-2.2-i2v-a14b" => {
            let img = p
                .image_data_uri
                .as_ref()
                .ok_or_else(|| anyhow!("Wan I2V requires a first-frame image"))?;
            let mut inp = json!({
                "prompt": p.prompt,
                "image": img,
                "resolution": if p.resolution.is_empty() { "480p".into() } else { p.resolution.clone() },
                "num_frames": ex_i(ex, "num_frames", 81).min(100),
                "frames_per_second": ex_i(ex, "frames_per_second", 16).min(24),
                "sample_steps": ex_i(ex, "sample_steps", 40),
                "sample_shift": ex_i(ex, "sample_shift", 5),
                "go_fast": ex_b(ex, "go_fast", false),
            });
            if let Some(np) = ex_s(ex, "negative_prompt") {
                inp["negative_prompt"] = json!(np);
            }
            Ok(inp)
        }
        other => Err(anyhow!("no input builder for Replicate model: {other}")),
    }
}

#[async_trait::async_trait]
impl Provider for ReplicateProvider {
    async fn generate(
        &self,
        client: &reqwest::Client,
        model_id: &str,
        params: &GenParams,
        progress: &(dyn Fn(f64) + Send + Sync),
    ) -> anyhow::Result<String> {
        let input = build_input(model_id, params)?;

        progress(0.05);
        let resp = client
            .post(format!("{API}/models/{model_id}/predictions"))
            .header("Authorization", format!("Token {}", self.key))
            .json(&json!({ "input": input }))
            .timeout(Duration::from_secs(30))
            .send()
            .await
            .context("Replicate submit request failed")?;

        if !resp.status().is_success() {
            let status = resp.status();
            let body = resp.text().await.unwrap_or_default();
            return Err(anyhow!("Replicate submit failed ({status}): {}", truncate(&body, 300)));
        }

        let pred: Value = resp.json().await.context("Replicate submit: bad JSON")?;
        let pred_id = pred
            .get("id")
            .and_then(|x| x.as_str())
            .ok_or_else(|| anyhow!("Replicate submit: no prediction id"))?
            .to_string();

        progress(0.15);
        let poll_url = format!("{API}/predictions/{pred_id}");
        let deadline = tokio::time::Instant::now() + Duration::from_secs(600);
        loop {
            if tokio::time::Instant::now() >= deadline {
                return Err(anyhow!("Replicate generation timed out after 600s (prediction {pred_id})"));
            }
            tokio::time::sleep(Duration::from_secs(5)).await;

            let r = client
                .get(&poll_url)
                .header("Authorization", format!("Token {}", self.key))
                .timeout(Duration::from_secs(30))
                .send()
                .await
                .context("Replicate poll request failed")?;
            let data: Value = r.json().await.context("Replicate poll: bad JSON")?;
            let status = data.get("status").and_then(|x| x.as_str()).unwrap_or("");

            match status {
                "succeeded" => {
                    progress(0.6);
                    return extract_output(&data);
                }
                "failed" | "canceled" => {
                    let err = data
                        .get("error")
                        .and_then(|x| x.as_str())
                        .unwrap_or("unknown error (no details from API)");
                    return Err(anyhow!("Replicate {status}: {}", truncate(err, 300)));
                }
                _ => {} // starting / processing — keep polling
            }
        }
    }
}

/// Replicate output is a string URL or a list of URLs — take the first.
fn extract_output(data: &Value) -> anyhow::Result<String> {
    match data.get("output") {
        Some(Value::String(s)) => Ok(s.clone()),
        Some(Value::Array(arr)) => arr
            .first()
            .and_then(|x| x.as_str())
            .map(|s| s.to_string())
            .ok_or_else(|| anyhow!("Replicate: empty output array")),
        other => Err(anyhow!("Replicate: unexpected output shape: {other:?}")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn params(image: Option<&str>) -> GenParams {
        GenParams {
            prompt: "a neon city".into(),
            aspect_ratio: "9:16".into(),
            resolution: "768p".into(),
            duration: 10,
            image_data_uri: image.map(|s| s.to_string()),
            extra: Default::default(),
        }
    }

    fn params_with(image: Option<&str>, extra: Value) -> GenParams {
        let mut p = params(image);
        p.extra = extra.as_object().cloned().unwrap_or_default();
        p
    }

    #[test]
    fn wan_t2v_defaults_when_no_extra() {
        let inp = build_input("wan-video/wan-2.2-t2v-fast", &params(None)).unwrap();
        assert_eq!(inp["num_frames"], 81);
        assert_eq!(inp["sample_shift"], 12);
        assert_eq!(inp["go_fast"], true);
        // No negative_prompt / lora keys when unset.
        assert!(inp.get("negative_prompt").is_none());
        assert!(inp.get("lora_weights_transformer").is_none());
    }

    #[test]
    fn wan_t2v_threads_advanced_params() {
        // This is the whole parity fix: frontend knobs must reach the input.
        let inp = build_input(
            "wan-video/wan-2.2-t2v-fast",
            &params_with(
                None,
                json!({
                    "num_frames": 121,
                    "sample_shift": 8,
                    "go_fast": false,
                    "negative_prompt": "morphing, 123 text",
                    "lora_weights_transformer": "https://hf/lora.safetensors",
                    "lora_scale_transformer": 0.8
                }),
            ),
        )
        .unwrap();
        assert_eq!(inp["num_frames"], 121);
        assert_eq!(inp["sample_shift"], 8);
        assert_eq!(inp["go_fast"], false);
        // A numeric-looking negative prompt must stay a STRING, not coerce to a number.
        assert_eq!(inp["negative_prompt"], "morphing, 123 text");
        assert_eq!(inp["lora_weights_transformer"], "https://hf/lora.safetensors");
        assert_eq!(inp["lora_scale_transformer"], 0.8);
    }

    #[test]
    fn wan_i2v_caps_num_frames_at_100_and_fps_at_24() {
        let inp = build_input(
            "wan-video/wan-2.2-i2v-a14b",
            &params_with(Some("data:image/png;base64,AAAA"), json!({ "num_frames": 121, "frames_per_second": 30 })),
        )
        .unwrap();
        assert_eq!(inp["num_frames"], 100);
        assert_eq!(inp["frames_per_second"], 24);
    }

    #[test]
    fn hailuo_optimize_prompt_toggle_threads() {
        let inp = build_input(
            "minimax/hailuo-2.3",
            &params_with(None, json!({ "optimize_prompt": false })),
        )
        .unwrap();
        assert_eq!(inp["prompt_optimizer"], false);
    }

    #[test]
    fn hailuo_snaps_duration_and_1080_forces_6() {
        let mut p = params(None);
        p.resolution = "1080p".into();
        p.duration = 10;
        let inp = build_input("minimax/hailuo-2.3", &p).unwrap();
        // 1080p must force duration back to 6.
        assert_eq!(inp["duration"], 6);
        assert_eq!(inp["resolution"], "1080p");
    }

    #[test]
    fn hailuo_duration_snaps_to_6_when_low() {
        let mut p = params(None);
        p.resolution = "768p".into();
        p.duration = 4;
        let inp = build_input("minimax/hailuo-2.3", &p).unwrap();
        assert_eq!(inp["duration"], 6);
    }

    #[test]
    fn wan_i2v_requires_image() {
        let err = build_input("wan-video/wan-2.2-i2v-a14b", &params(None)).unwrap_err();
        assert!(err.to_string().contains("requires a first-frame image"));
        // With an image it builds and includes it.
        let ok = build_input("wan-video/wan-2.2-i2v-a14b", &params(Some("data:image/png;base64,AAAA"))).unwrap();
        assert_eq!(ok["image"], "data:image/png;base64,AAAA");
    }

    #[test]
    fn unknown_model_rejected() {
        assert!(build_input("some/unwired-model", &params(None)).is_err());
    }

    #[test]
    fn extract_output_handles_shapes() {
        assert_eq!(
            extract_output(&json!({ "output": "https://x/v.mp4" })).unwrap(),
            "https://x/v.mp4"
        );
        assert_eq!(
            extract_output(&json!({ "output": ["https://x/a.mp4", "https://x/b.mp4"] })).unwrap(),
            "https://x/a.mp4"
        );
        assert!(extract_output(&json!({ "output": [] })).is_err());
        assert!(extract_output(&json!({ "output": 42 })).is_err());
        assert!(extract_output(&json!({})).is_err());
    }
}
